### PR TITLE
feat(auth): create new user

### PR DESCRIPTION
**Pull request by bhaskarhc**
#### Test Description :

 Create **new user** on Kubera on-prem by sending a post request to its API server

#### Endpoint :
 ```
http://<Kubera_Host>/v3/localauthconfig/
```
#### Body :
``` json
{
    "firstName": "<user_first_name>",
    "lastName": "<user_last_name>",
    "password": {
        "publicValue": "<user_email>",
        "secretValue": "<user_password>"
    }
}
```
#### Folder Structure

```
|_ kubera-api-testing/
    |__ kubera/
    |   |__ connection/
    |   |    |__ connection.go
    |   |__ authentication/
    |        |__ authentication_test.go
    |        |__ authentication.go
    |        |__ user_create_test.go
    |        |__ user_create.go
    |__ go.mod
    |__ go.sum
    |__ makefile

```

Test Scenarios :

```json
{
		"No host": {
			FirstName: "test",
			LastName:  "testF",
			Email:     "test@test.io",
			Password:  "P@ssw0rd",
			IsError:   true,
		},
		"No firstName": {
			HostName: HostName,
			LastName: "testF",
			Email:    "test@test.io",
			Password: "P@ssw0rd",
			IsError:  true,
		},
		"No Email": {
			HostName:  HostName,
			FirstName: "test",
			LastName:  "testF",
			Password:  "P@ssw0rd",
			IsError:   true,
		},
		"With No password": {
			HostName:  HostName,
			FirstName: "test",
			LastName:  "testF",
			Email:     "test@test.io",
			IsError:   true,
		},
		"With host, firstName, lastName, email & password": {
			HostName:  connection.HostName,
			FirstName: "test",
			LastName:  "testF",
			Email:     "test@test.io", 
			Password:  "P@ssw0rd",
		},
		"With invalid email": {
			HostName:  connection.HostName,
			FirstName: "test",
			LastName:  "testF",
			Email:     "testtest.io",
			Password:  "P@ssw0rd",
			IsError:   true,
		},
	}
```

**Expected output**
```javascript
bhaskarjb@Bhaskars-MBP:~/Desktop/mayadata-io/kubera-api-testing|signup-testcase⚡ 
⇒  go test  ./...                                                                                                         
ok  	github.com/mayadata-io/kubera-api-testing/kubera/authentication	1.739s
?   	github.com/mayadata-io/kubera-api-testing/kubera/connection	[no test files]
```

**Test case fail output**
```javascript
bhaskarjb@Bhaskars-MBP:~/Desktop/mayadata-io/kubera-api-testing|signup-testcase⚡ 
⇒  go test  ./...     
--- FAIL: TestCreateNewUser (0.73s)
    --- FAIL: TestCreateNewUser/No_lastName (0.10s)
        create_user_test.go:99: Expected no error got={
             ."err": null,
             ."message": "Please use a different email as this email is already taken",
             ."statusCode": 409,
             ."status": "409 Conflict"
             }: 
            Creator={
             ."firstName": "test",
             ."lastName": "",
             ."password": {
             .."publicValue": "nolastname122@test.io",
             .."secretValue": "XXXX"
             .}
             }: 
            Token={
             ."jwt": "",
             ."accountId": ""
             }
    --- FAIL: TestCreateNewUser/With_firstName,_lastName,_email_&_password (0.07s)
        create_user_test.go:99: Expected no error got={
             ."err": null,
             ."message": "Please use a different email as this email is already taken",
             ."statusCode": 409,
             ."status": "409 Conflict"
             }: 
            Creator={
             ."firstName": "test",
             ."lastName": "testF",
             ."password": {
             .."publicValue": "testing12333@test.io",
             .."secretValue": "XXXX"
             .}
             }: 
            Token={
             ."jwt": "",
             ."accountId": ""
             }
FAIL
FAIL	github.com/mayadata-io/kubera-api-testing/kubera/authentication	1.425s
?   	github.com/mayadata-io/kubera-api-testing/kubera/connection	[no test files]
```

### CODE REVIEW

File: kubera/authentication/authentication_test.go [hunk 1]
```diff
@@ -75,3 +75,91 @@ func TestFetchToken(t *testing.T) {
...
+			IsError:  true,
+		},
+		// Without Lastname also create user
+		// "No lastName": {
+		// 	HostName:  connection.HostName,
+		// 	FirstName: "test",
+		// 	Email:     "test@test.io",
+		// 	Password:  "P@ssw0rd",
+		// 	IsError:   true,
+		// },
```
Comment (amitbhatt818) [inline kubera/authentication/authentication_test.go]: If it is creating new user without `lastName` then you can convert this to positive scenario. @bhaskarhc

**Review by amitbhatt818 — CHANGES_REQUESTED** [review]

File: kubera/authentication/authentication_test.go [hunk 2]
```diff
@@ -75,3 +75,89 @@ func TestFetchToken(t *testing.T) {
 		})
 	}
 }
+
+// TestCreateNewUser is to create a new user
+func TestCreateNewUser(t *testing.T) {
```
Comment (AmitKumarDas) [inline kubera/authentication/authentication_test.go]: Let us have this in a separate _test.go file e.g. authentication/create_user_test.go

File: kubera/authentication/create_user.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
+package authentication
+
+import (
+	"encoding/json"
+
+	"github.com/go-resty/resty/v2"
+	conn "github.com/mayadata-io/kubera-api-testing/kubera/connection"
+)
+
+type NewUserForm struct {
```
Comment (AmitKumarDas) [inline kubera/authentication/create_user.go]:
```go
type User struct {}
 
 type UserCreationConfig struct {
   // fields required to create a new instance of UserCreation
   // e.g credentials
 }
 
 type UserCreation struct {
   // input to rest api
   // e.g. credentials
 }

 func NewUserCreator(config UserCreationConfig) *UserCreation {
   return &UserCreation{}
 }
 
 func (u *UserCreation) Create() (User, error) {
   // rest api call
 }
```

Comment (AmitKumarDas) [reply kubera/authentication/create_user.go]:
```go
  type <Payload> struct {
   // fields returned by Kubera
   // json hints
  }

  type <Input-ing/ionConfig> struct {
   // mostly same fields to that of <Input-ing>
  }

  // TokenFetching
  // UserCreation
  type <Input-ing/ion> struct {
    // fields that will be used to send REST API
  }
  
  func Constructor(config Input-ingConfig) (<Payload>, error) {
   // REST API call
  }
```

File: kubera/authentication/create_user.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	resp, err := client.R().
+		SetHeader("Content-Type", "application/json").
+		SetBody(credJSON).
+		//SetError method automatic unmarshalling if response status code >= 399
+		// and content type either JSON or XML.
+		SetError(&kErr).
+		//SetResult automatic unmarshalling for the request,
+		// if response status code is between 200 and 299
+		SetResult(&token).
+		Post(url)
```
Comment (rohan2794) [inline kubera/authentication/create_user.go:76]: Avoid extra `url `variable
Post(fmt.Sprintf("%sv3/localauthconfig/", connection.Get().HostName))

File: kubera/authentication/create_user.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
+package authentication
+
+import (
+	"encoding/json"
+
+	"github.com/go-resty/resty/v2"
+	conn "github.com/mayadata-io/kubera-api-testing/kubera/connection"
```
Comment (rohan2794) [inline kubera/authentication/create_user.go]: avoid alias `conn` , we can directly refer it by `connection` as it is used at only one place in whole file.

File: kubera/authentication/create_user.go (hunk 4 above)
Comment (bhaskarhc) [reply kubera/authentication/create_user.go:76]: hv plan to create a hostname check function in connection package , which will return hostname as `http://xx.xx.xx.xxx` and all endpoint will start with `/` .

File: kubera/authentication/create_user.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+// UserCreation contains structure of the input data to create NewUser
+type UserCreation struct {
+	FirstName   string   `json:"firstName"`
+	LastName    string   `json:"lastName"`
+	Credentials UserCred `json:"password"`
+}
+
+// UserCred contains structure of the user secrets input
+type UserCred struct {
```
Comment (AmitKumarDas) [inline kubera/authentication/create_user.go]: `UserCredential`

File: kubera/authentication/create_user.go [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+	"github.com/go-resty/resty/v2"
+	"github.com/mayadata-io/kubera-api-testing/kubera/connection"
+)
+
+// UserCreation contains structure of the input data to create NewUser
+type UserCreation struct {
+	FirstName   string   `json:"firstName"`
+	LastName    string   `json:"lastName"`
+	Credentials UserCred `json:"password"`
```
Comment (AmitKumarDas) [inline kubera/authentication/create_user.go]: Credential UserCredential `json:"password"`

File: kubera/authentication/create_user.go [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+// Create is  a method , It creates new user in kubera by making -
+//  post reqest to API server
+func (t *UserCreation) Create() (Token, error) {
+	client := resty.New()
+	url := connection.Get().HostName + "v3/localauthconfig/"
+	token := Token{}
+	kErr := KuberaError{}
+	credJSON, err := json.Marshal(t)
```
Comment (AmitKumarDas) [inline kubera/authentication/create_user.go]: missing error check

File: kubera/authentication/create_user.go [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+			Password: config.Credentials.Password,
+		},
+	}
+}
+
+// Create is  a method , It creates new user in kubera by making -
+//  post reqest to API server
+func (t *UserCreation) Create() (Token, error) {
+	client := resty.New()
+	url := connection.Get().HostName + "v3/localauthconfig/"
```
Comment (AmitKumarDas) [inline kubera/authentication/create_user.go]: make use of fmt.Sprintf

File: kubera/authentication/create_user.go (hunk 8 above)
Comment (AmitKumarDas) [inline kubera/authentication/create_user.go]: rename credJson to body

File: kubera/authentication/create_user_test.go [hunk 10]
```diff
@@ -0,0 +1,78 @@
+package authentication
+
+import (
+	"testing"
+)
+
+// TestCreateNewUser is to create a new user
```
Comment (AmitKumarDas) [inline kubera/authentication/create_user_test.go]: comments need to be reworked

File: kubera/authentication/create_user.go [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+	token := Token{}
```
Comment (AmitKumarDas) [inline kubera/authentication/create_user.go]: rename `token` to `response`

File: kubera/authentication/create_user_test.go [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+				UserCreation: UserCreation{
+					FirstName: mock.FirstName,
+					LastName:  mock.LastName,
+					Credentials: UserCred{
+						Email:    mock.Email,
+						Password: mock.Password,
+					},
+				},
+			})
+			createUser, err := creator.Create()
```
Comment (AmitKumarDas) [inline kubera/authentication/create_user_test.go]: rename `createUser` to `got`

File: kubera/authentication/create_user.go [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+			Password: config.Credential.Password,
+		},
+	}
+}
+
+// Create is  a method , It creates new user in kubera by making -
+//  post reqest to API server
+func (t *UserCreation) Create() (Token, error) {
+	client := resty.New()
+	// url := connection.Get().HostName + "v3/localauthconfig/"
```
Comment (AmitKumarDas) [inline kubera/authentication/create_user.go]: remove comment

File: kubera/authentication/create_user.go [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+	url := fmt.Sprintf("%s/v3/localauthconfig/", connection.Get().HostName)
```
Comment (AmitKumarDas) [inline kubera/authentication/create_user.go]: localauthconfig can be some field within UserCreation struct.
In other words, how do we test google auth, github auth & so on.
When are we going to test google/github based auths?

File: kubera/authentication/create_user.go [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+		FirstName: config.FirstName,
+		LastName:  config.LastName,
+		Credential: UserCredential{
+			Email:    config.Credential.Email,
+			Password: config.Credential.Password,
+		},
+	}
+}
+
+// Create will creates new kubera user in kubera by making post reqest
```
Comment (harshshekhar15) [inline kubera/authentication/create_user.go]: typo - `request`

File: kubera/authentication/create_user.go (hunk 14 above)
Comment (bhaskarhc) [reply kubera/authentication/create_user.go]: Google and Github auth have totally different kind of body and endpoint .

File: kubera/authentication/create_user.go (hunk 15 above)
Comment (bhaskarhc) [reply kubera/authentication/create_user.go]: done

File: kubera/authentication/create_user.go (hunk 13 above)
Comment (bhaskarhc) [reply kubera/authentication/create_user.go]: done

File: kubera/authentication/create_user.go (hunk 5 above)
Comment (bhaskarhc) [reply kubera/authentication/create_user.go]: done

File: kubera/authentication/create_user.go (hunk 4 above)
Comment (bhaskarhc) [reply kubera/authentication/create_user.go:76]: done

**Review by rohan2794 — APPROVED** [review]
lgtm

**Review by harshshekhar15 — APPROVED** [review]
/lgtm

**Review by amitbhatt818 — APPROVED** [review]
/lgtm

**Review by AmitKumarDas — APPROVED** [review]

**Pull request merged**